File: prediction_generation/cpdbench_prophet.py
import argparse
import time
import pandas as pd
from prophet import Prophet
import json
import sys
import copy

from cpdbench_utils import load_dataset, exit_success, make_param_dict, exit_with_error

def parse_args():
    parser = argparse.ArgumentParser(description="Wrapper for Prophet")
    parser.add_argument("-i", "--input", help="path to the input data file", required=True)
    parser.add_argument("-o", "--output", help="path to the output file")
    parser.add_argument("-N", "--Nmax", help="maximum number of changepoints", choices=['default', 'max'])
    parser.add_argument("-w", "--WeeklySeasonality", type=bool, help="Weekly Seasonality")
    parser.add_argument("-d", "--DailySeasonality", type=bool, help="Daily Seasonality")
    parser.add_argument("-r", "--ChangepointRange", type=float, help="Changepoint Range")
    parser.add_argument("-p", "--ChangepointPriorScale", type=float, help="Changepoint Prior Scale")
    parser.add_argument("-t", "--IntervalWidth", type=float, help="Interval Width")
    parser.add_argument("-g", "--growth", type=str, help="Growth type: 'linear' or 'logistic'", choices=['linear', 'logistic'])
    parser.add_argument("-c", "--cap", type=float, help="Capacity for logistic growth (required for logistic growth)")

    return parser.parse_args()

# Function to convert Timestamp objects to a JSON serializable format
def convert_timestamps(obj):
    if isinstance(obj, pd.Timestamp):
        return obj.isoformat()  # Convert to ISO 8601 string format
    elif isinstance(obj, list):
        return [convert_timestamps(item) for item in obj]
    elif isinstance(obj, dict):
        return {key: convert_timestamps(value) for key, value in obj.items()}
    else:
        return obj

def main():
    args = parse_args()
    raw_args = copy.deepcopy(args)

    # Load the dataset (using a Python equivalent of your R helper function)
    data, mat = load_dataset(args.input)

    start_time = time.time()

    if args.Nmax == 'default':
        args.Nmax = 25
    else:
        args.Nmax = data['n_obs'] - 1  # Adjusted from 'original' to the main data structure

    # Check if 'series' is in data and extract appropriately
    if 'series' not in data or len(data['series']) == 0:
        exit_with_error(data, raw_args, vars(args), "No time series data available.")

    # Prepare the DataFrame for Prophet
    df = pd.DataFrame({
        'ds': data['time']['raw'],  # Time column
        'y': data['series'][0]['raw']  # Series values
    })

    # Handle logistic growth by adding 'cap'
    if args.growth == 'logistic':
        if args.cap is None:
            exit_with_error(data, raw_args, vars(args), "Capacity ('cap') must be provided for logistic growth.")
        
        # Add the capacity column to the DataFrame
        df['cap'] = args.cap

    # Fit the Prophet model
    try:
        model = Prophet(
            changepoint_range=args.ChangepointRange,
            n_changepoints=args.Nmax,
            weekly_seasonality=args.WeeklySeasonality,
            daily_seasonality=args.DailySeasonality,
            growth=args.growth,  # 'linear' or 'logistic'
            changepoint_prior_scale=args.ChangepointPriorScale,
            interval_width=args.IntervalWidth
        )
        model.fit(df)

        # Retrieve changepoints (timestamps)
        changepoints = model.changepoints
        locs = changepoints.index.tolist()

    except Exception as e:
        exit_with_error(data, raw_args, vars(args), str(e), __file__)

    stop_time = time.time()
    runtime = stop_time - start_time

    locs = convert_timestamps(locs)
    data = convert_timestamps(data)

    exit_success(data, raw_args, vars(args), locs, runtime, __file__)

if __name__ == "__main__":
    main()

File: prediction_generation/old-code/cpdbench_prophet.R
#' ---
#' title: Wrapper for the Prophet package in TCPDBench
#' author: G.J.J. van den Burg
#' date: 2019-09-30
#' license: See the LICENSE file.
#' copyright: 2019, The Alan Turing Institute
#' ---

Sys.setenv(TZ="UTC")
library(argparse)
library(prophet)
library(lubridate)

load.utils <- function() {
    # get the name of the current script so we can load utils.R (yay, R!)
    cmd.args <- commandArgs(trailingOnly=F)
    file.arg <- "--file="
    this.script <- sub(file.arg, "", cmd.args[grep(file.arg, cmd.args)])
    this.dir <- dirname(this.script)
    utils.script <- file.path(this.dir, 'utils.R')
    source(utils.script)
}

parse.args <- function() {
    parser <- ArgumentParser(description='Wrapper for Prophet package')
    parser$add_argument('-i',
                        '--input',
                        help='path to the input data file',
                        required=TRUE
    )
    parser$add_argument('-o',
                        '--output',
                        help='path to the output file'
    )
    parser$add_argument('-N',
                        '--Nmax',
                        help='maximum number of changepoints',
                        choices=c('default', 'max')
    )
    return(parser$parse_args())
}

frac.to.dt <- function(raw) {
    out <- c()
    for (i in 1:length(raw)) {
        replaced <- gsub('-', '.', raw[i]);
        number <- as.double(replaced)
        year <- floor(number)
        remainder <- number - year
        begin <- as_datetime(paste(year, '-01-01', sep=''))
        end <- as_datetime(paste(year+1, '-01-01', sep=''))
        offset <- remainder * (end - begin)
        dt <- begin + offset
        # you'd think there'd be a well-documented easy-to-find function for
        # this
        datepart <- date(dt)
        timepart <- sprintf("%02d:%02d:%02d", hour(dt), minute(dt),
                            round(second(dt)))
        iso <- paste(datepart, timepart, sep=' ')
        out <- c(out, iso)
    }
    return(out)
}

preprocess.data <- function(data)
{
    if (data$original$time$format == "%Y-%m-%d %H:%M:%S") {
        tidx <- data$original$time$raw
    } else if (data$original$time$format == "%Y-%m-%d") {
        tidx <- data$original$time$raw
    } else if (data$original$time$format == "%Y-%m") {
        tidx <- paste(data$original$time$raw, '-01', sep='')
    } else if (data$original$time$format == "%Y") {
        tidx <- paste(data$original$time$raw, '-01-01', sep='')
    } else if (data$original$time$format == "%Y-%F") {
        tidx <- frac.to.dt(data$original$time$raw)
    } else {
        stop(cat("Unknown time format: ", data$original$time$format, '\n'))
    }

    raw <- as.vector(data$mat)

    df <- data.frame(ds=tidx, y=raw)

    return(df)
}

main <- function() {
    args <- parse.args()
    data <- load.dataset(args$input)

    defaults <- list()
    # we want to allow change points throughout the entire range of the series
    defaults$changepoint.range <- 1
    # threshold used in add_changepoints_to_plot
    defaults$threshold <- 0.01
    defaults$yearly.seasonality <- 'auto'
    defaults$weekly.seasonality <- 'auto'
    defaults$daily.seasonality <- 'auto'

    if (args$Nmax == 'default')
        args$Nmax <- 25
    else
        args$Nmax <- data$original$n_obs - 1

    if (!(data$original$name %in% c('bank', 'bee_waggle'))) {
        defaults$yearly.seasonality <- FALSE
        defaults$weekly.seasonality <- FALSE
        defaults$daily.seasonality <- FALSE
    }

    params <- make.param.list(args, defaults)

    if (data$original$n_dim > 1) {
        # package doesn't handle multidimensional data
        exit.error.multidim(data$original, args, params)
    }

    df <- preprocess.data(data)

    start.time <- Sys.time()
    result <- tryCatch({
        model <- prophet(df, changepoint.range=params$changepoint.range,
                         n.changepoints=params$Nmax,
                         yearly.seasonality=params$yearly.seasonality,
                         weekly.seasonality=params$weekly.seasonality,
                         daily.seasonality=params$daily.seasonality
        )
        threshold <- params$threshold
        cpt <- model$changepoints[abs(model$params$delta) >= threshold]
        cpt <- as.character(as.POSIXct(cpt))
        locs <- match(cpt, as.character(df$ds))
        list(locations=locs, error=NULL)
    }, error=function(e) {
        return(list(locations=NULL, error=e$message))
    })
    stop.time <- Sys.time()
    runtime <- difftime(stop.time, start.time, units='secs')

    if (!is.null(result$error))
        exit.with.error(data$original, args, params, result$error)

    # convert to 0-based indices
    locations <- as.list(result$locations)

    exit.success(data$original, args, params, locations, runtime)
}

load.utils()
main()